### PR TITLE
Rework lsp.util.apply_text_edits to use nvim_buf_set_text

### PR DESCRIPTION
**Pull request by Jesse-Bakker**
Among other things, fixes cursor jumping for some textDocument/formatting responses and retains extmarks

cc @mfussenegger: not sure if this supersedes or is complementary to #13677

### CODE REVIEW

**Comment by runiq** [comment]
Cribbed from [chat](https://gitter.im/neovim/neovim?at=5ffdf21514cec811ec944490):

I thought with `nvim_buf_set_text` being merged, we could simply apply LSP `TextEdit`s one by one, but that would send one `textDocument/didChange` notification *for each* `TextEdit` and be against the spec (where all `TextEdit`s *together* are supposed to move a document from state S1 to S2).

To work around this, we use `vim.lsp.util.set_lines` to apply the `TextEdit`s to a *copy* of the buffer, and then set the actual new document text using that copy. That's fairly roundabout. :(

@bfredl Is there a way to 'opt out' of the `textDocument/didChange` notification until all `TextEdit`s have been applied to the buffer?

**Comment by bfredl** [comment]
@runiq I think this can be done in user space simply by LSP module internally having a variable to inhibit/buffer on_bytes calls when caused by didChange.

**Comment by Jesse-Bakker** [comment]
Imo the edits should also be atomic wrt undo, meaning applying a series of textedits should be undone with a single undo

**Comment by tjdevries** [comment]
`nvim_atomic_undo(lua_function_here)`? Is it possible @bfredl ? Is `nvim_call_atomic` already this (but it's not exposed internally).

Idk, just thinking out loud here.

**Comment by Jesse-Bakker** [comment]
AFAICT `nvim_call_atomic` just batches RPC calls.

Making the undo atomic should not be too hard, as long as the edits don't overlap and don't execute undo/redo's (save the pointer to the undo head and after executing all edits walk the new undo entries and concatenate their u_entry lists). The problem there is that autocommands might be fired for every intermediate undo, which will make the autocommands get out of sync with the edited undo tree. I do not know if it is possible to inhibit these and just call one autocommand at the end.

**Comment by tjdevries** [comment]
@Jesse-Bakker we could do the edits with `:noautocmd` and then apply them after... but that doesn't feel right :laughing:

**Comment by Jesse-Bakker** [comment]
The cleanest might be something like `nvim_buf_apply_edits(bufnr, edits)`, where edits is a list containing tables with `nvim_buf_set_text`s remaining arguments. That can be a generalisation of `nvim_buf_set_text`, which can then be implemented on top of that.

**Comment by Jesse-Bakker** [comment]
Things that from my view would need to happen to support applying multiple edits atomically:

 - Change on_lines callbacks to receive a list of changed blocks
 - Implement nvim_buf_apply_edits API

For implementing the API, we need to keep in mind that, internally, edits are all line based, so two edits on the same line could create problems

In the end making these changes would result in us sending much smaller diffs to language servers (instead of sending a change of all lines between the "top" and "bottom" edit, we only send changed lines).

**Comment by bfredl** [comment]
Hmm, ideally updates should be batched automatically whenever they form the same undo header. this way, the callback behaviour would be consistent between undo and redo as well.

**Comment by runiq** [comment]
`nvim_buf_lines_event` has the `more` flag to indicate that this is part of a chunked change; could this be useful somehow?

**Comment by Jesse-Bakker** [comment]
So, getting back into this:

> `nvim_buf_lines_event` has the `more` flag to indicate that this is part of a chunked change; could this be useful somehow?

This event is used externally, where the `more` flag indicates that the diff is too large for a single message. Internally, in the `on_lines` callback, we do not have such a flag.

So as far as I could find, there is currently no way to have a callback for changes to multiple disconnected lines.

> Hmm, ideally updates should be batched automatically whenever they form the same undo header. this way, the callback behaviour would be consistent between undo and redo as well.

This would be awesome to have, as callbacks could just apply a series of edits and not have to worry about batching them using an API function or anything.

In the end the change I would propose is a new event (`on_change`/`on_changes`?), which gets a list of changes instead of a single change. This event would be emitted after finishing the undo header (so after a function call finishes).

**Comment by bfredl** [comment]
> In the end the change I would propose is a new event (on_change/on_changes?), which gets a list of changes instead of a single change.

Hmm I guess it depends how big we expect a typical batch edit to be. Having 10 function calls instead of one is not a big deal (stil batched of course, to merge overlapping edits), but it gets up to 100-1000 it is maybe a different story.

**Comment by Jesse-Bakker** [comment]
I think the current state of this PR is quite nice in that it improves the user facing behavior, while it should not perform worse. Getting to the ideal state discussed here is maybe something for after 0.5, in a different PR, as I do not really see it happening before then. So I'd suggest reviewing this as-is and continuing discussion in an issue. WDYT?

File: runtime/lua/vim/lsp/util.lua [hunk 1]
```diff
@@ -760,9 +817,9 @@ function M.apply_workspace_edit(workspace_edit)
     for idx, change in ipairs(workspace_edit.documentChanges) do
       if change.kind == "rename" then
         M.rename(
-          vim.uri_to_fname(change.oldUri),
```
Comment (mfussenegger) [inline runtime/lua/vim/lsp/util.lua:763]: Could you revert the unrelated formatting changes?

**Review by mfussenegger — COMMENTED** [review]
@mjlbach this might even be something to get into 0.5.0 - I think it would help preventing text edits from messing up extmarks.

File: runtime/lua/vim/lsp/util.lua (hunk 1 above)
Comment (Jesse-Bakker) [reply runtime/lua/vim/lsp/util.lua]: Oops, format-on-save getting me once again :sweat_smile:

**Comment by mjlbach** [comment]
Linux lint errors need to be fixed still.

@bfredl, I'm fine with merging this. Are you satisfied with the changes?

**Comment by runiq** [comment]
> this might even be something to get into 0.5.0 - I think it would help preventing text edits from messing up extmarks.

@mfussenegger Just to make sure: With the approach in this PR, extmarks that happen to be between two text edits will still get clobbered, right?

**Comment by clason** [comment]
bump? still on track for 0.5?

**Comment by folke** [comment]
I'm currently debugging an issue with diagnostics virtual text not updating properly on buffer updates. I tried this PR to see if it fixes my issue, but with this it's actually worse.

All virtual text from diagnostics is moved to the last line. Without this PR, the virtual text is removed.

Running `vim.lsp.diagnostic.display(nil, bufnr)`, then restores the virtual text, but this should not be needed, since the LSP client listens for on_lines events and will schedule a diagnostics check.

The root cause of the issue I'm trying to solve is related to nvim_buf_attach, where on_lines seems to get called too early in an async context. So any plugin (or LSP client in this case, but also one of my plugins todo-comments) that does something in on_lines, will do this too early and those changes will again be overwritten. For todo-comments I have a hack in place that defers the update in on_lines.

I'm further looking into this, and will open an issue once I have a better idea what is happening and causing this.

For anyone interested, the issue can be triggered by `luafile %`:

```lua
print(foo)

local lines = vim.api.nvim_buf_get_lines(0, 0, -1, false)
vim.api.nvim_buf_set_lines(0, 0, -1, false, lines)
-- vim.api.nvim_buf_set_text(0, 0, 0, #lines - 1, #lines[#lines], lines)
```

Wait till the lua lsp shows a warning for undefined `foo`.

Then run the file with either set_lines or set_text.

Both should actually retain the virtual text, but because of that on_lines being called too early, it either gets removed, or gets moved to the last line.

**Comment by Jesse-Bakker** [comment]
A more easily reproducible example, without using any lsp, is:

```lua
local ns = vim.api.nvim_create_namespace("testns")
local extmark = vim.api.nvim_buf_set_extmark(0, ns, 0, 0, {virt_text = {{"test virttext"}}})

local lines = vim.api.nvim_buf_get_lines(0, 0, -1, false)
vim.api.nvim_buf_set_text(0, 0, 0, #lines - 1, #(lines[#lines]), lines)
```

You can see the same happening, with the virtual text extmark shifting to the last line. Looks like a bug with `nvim_buf_set_text` incorrectly retaining extmarks on unchanged lines, instead shifting them to the end.

**Comment by folke** [comment]
@Jesse-Bakker so if I understand correctly, the "preserving" only means that in case of set_lines it will be removed, but in case of `set_text` it will be moved to the end of the replaced text?

So basically only when doing `nvim_buf_set_lines` line per line, only then the extmarks would be preserved.

But that would trigger multiple document updates to lsp servers, which is not what we want.

**Comment by Jesse-Bakker** [comment]
Perfect preservation of extmarks is very difficult within lsp edits, because extmarks attach to locations, not semantics. The only "correct" thing to do for diagnostics is to recalculate them on the state of the file after applying the edits.

`set_text` is useful because it can preserve extmarks on a part of a line you do not touch. For example:
```
asdf
 ^
```
and replace `d->g` (`nvim_buf_set_text(0, 0, 2, 0, 3, {"g"})`)
```
asdg
 ^
```
whereas with `set_lines` that extmark would disappear.

So ideally, we would be able to `set_text` for individual edits and only emit one change event, as noted before in this PR.

**Comment by folke** [comment]
@Jesse-Bakker there was an issue where diagnostics virtual text got removed. See #15011 

There, we restore any extmarks that are no longer in the buffer.

That works well with the current `apply_text_edits` since `nvim_buf_set_lines` removes the actual extmarks.

However, with `nvim_buf_set_text`, the extmarks will me moved to the last line (or last line of the last edit?).

In this case, I won't touch those extmarks since they are still valid.

Any ideas on how to resolve that?

Should this PR explicitely remove diagnostics for the edit ranges?

**Comment by folke** [comment]
I just created a follow-up PR to further fix the issue with diagnostics virtual text.

This PR will still break it again, but there's a small fix that can be included here to fix it again. See #15023

So probably best to wait till #15023 is merged, then rebase on master and include the fix mentioned.

**Comment by craftzdog** [comment]
Looks like #15023 is already merged. Could you consider merging it, please? 🙏

**Comment by mfussenegger** [comment]
I think this one is obsolete, now that https://github.com/neovim/neovim/pull/15561 got merged, right?
